src/pool/srv_pool_check.c

```diff
@@ -307,3 +307,471 @@ ds_pool_clues_print(struct ds_pool_clues *clues)
 			dc->bcl_oid_next);
 	}
 }
+
+int
+ds_pool_clues_find_rank(struct ds_pool_clues *clues, d_rank_t rank)
+{
+	int i;
+
+	for (i = 0; i < clues->pcs_len; i++)
+		if (clues->pcs_array[i].pc_rank == rank)
+			return i;
+	return -DER_NONEXIST;
+}
+
+/*
+ * Return
+ *
+ *   > 0	<x_last_term, x_last_index> is newer than <y_last_term, y_last_index>
+ *   < 0	<x_last_term, x_last_index> is older than <y_last_term, y_last_index>
+ *   = 0	<x_last_term, x_last_index> is equal to   <y_last_term, y_last_index>
+ */
+static int
+compare_logs(uint64_t x_last_term, uint64_t x_last_index,
+	     uint64_t y_last_term, uint64_t y_last_index)
+{
+	if (x_last_term > y_last_term)
+		return 1;
+	if (x_last_term < y_last_term)
+		return -1;
+
+	if (x_last_index > y_last_index)
+		return 1;
+	if (x_last_index < y_last_index)
+		return -1;
+
+	return 0;
+}
+
+/**
+ * Analyze \a clues, which must be nonempty and comprise clues about replicas
+ * of one PS, and report if this PS requires catastrophic recovery or not.
+ *
+ * \param[in]	clues		pool clues for one PS
+ * \param[out]	advice_out	when the return value is 1, the index of the
+ *				advised replica in \a clues to rebootstrap the
+ *				PS from
+ *
+ * \return	0	this PS does not require catastrophic recovery
+ *		>0	the caller is advised to rebootstrap this PS from the
+ *			replica at index \a *advice_out in \a clues
+ */
+int
+ds_pool_check_svc_clues(struct ds_pool_clues *clues, int *advice_out)
+{
+	uuid_t		uuid;
+	uint32_t	map_version = 0;
+	uint64_t	log_term = 0;
+	uint64_t	log_index = 0;
+	int		advice = -1;
+	int		i;
+
+	/* Assert that all clues are about replicas of the same PS. */
+	D_ASSERT(clues->pcs_len > 0);
+	uuid_copy(uuid, clues->pcs_array[0].pc_uuid);
+	for (i = 0; i < clues->pcs_len; i++) {
+		struct ds_pool_clue *clue = &clues->pcs_array[i];
+
+		D_ASSERT(uuid_compare(uuid, clue->pc_uuid) == 0);
+		D_ASSERTF(clue->pc_rc == 0, DF_RC"\n", DP_RC(clue->pc_rc));
```

Comment:
I have updated "pc_rc" logic: use it to indicate whether pc_svc_clue exist or not when transfer the clue on wire. I can refresh the patch when merge with my patches.